File: 146-lru-cache/golang/solution.go
package golang

type Node struct {
	Key  int
	Val  int
	Prev *Node
	Next *Node
}

type LRUCache struct {
	capacity int
	cache    map[int]*Node
	head     *Node
	tail     *Node
}

func Constructor(capacity int) LRUCache {
	head := &Node{Key: -1, Val: -1, Prev: nil, Next: nil}
	tail := &Node{Key: -1, Val: -1, Prev: nil, Next: nil}
	head.Next = tail
	tail.Prev = head
	return LRUCache{
		capacity: capacity,
		cache:    make(map[int]*Node, capacity),
		head:     head,
		tail:     tail,
	}
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.cache[key]; ok {
		this.remove(node)
		this.add(node)
		return node.Val
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	node, ok := this.cache[key]
	if ok {
		this.remove(node)
	}
	node = &Node{
		Key:  key,
		Val:  value,
		Next: nil,
		Prev: nil,
	}
	this.add(node)
	this.cache[key] = node

	if len(this.cache) > this.capacity {
		next := this.head.Next
		this.remove(next)
		delete(this.cache, next.Key)
	}

}

func (this *LRUCache) remove(node *Node) {
	prev := node.Prev
	next := node.Next
	prev.Next = next
	next.Prev = prev
}

func (this *LRUCache) add(node *Node) {
	p := this.tail.Prev
	p.Next = node
	this.tail.Prev = node
	node.Prev = p
	node.Next = this.tail
}

File: 146-lru-cache/python/solution.py
from dataclasses import dataclass
from typing import Optional

@dataclass
class Node:
    key: int
    val: int
    prev: Optional["Node"] = None
    next: Optional["Node"] = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.dct = {}
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

        
    def get(self, key: int) -> int:
        if key in self.dct:
            node = self.dct[key]
            self._remove_node(node)
            self._add_node(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.dct:
            node = self.dct[key]
            self._remove_node(node)
            self._add_node(node)
            node.val = value
        else:     
            if len(self.dct) == self.capacity:
                node_to_del = self.head.next
                self._remove_node(node_to_del)
                del self.dct[node_to_del.key]
            node = Node(key=key, val=value)
            self._add_node(node)
            self.dct[key] = node

    def _remove_node(self, node: Node) -> None:
        prev = node.prev
        nxt = node.next
        prev.next = nxt
        nxt.prev = prev
        
    def _add_node(self, node: Node) -> None:
        last = self.tail.prev
        last.next = node
        self.tail.prev = node
        node.next = self.tail
        node.prev = last
